28 deploy to GitHub pages

matthewscottgordon opened the pull request:
Add GitHub Actions workflow to build and deploy to GitHub pages.

@tienhoah this is set up to build when there's a pull request and then deploy when it's merged to main. I'm guessing that's probably good for this project: main gets automatically deployed to a staging environment and deployment to production happens manually. Let me know if you want to do it differently.

@alexgleith I don't have the permissions needed to actually set up the deployment environment (`Branch "28-deploy-to-github-pages" is not allowed to deploy to github-pages due to environment protection rules.`) but I assume you can set that up and connect the custom subdomain?

matthewscottgordon committed:
Add *~ to .gitignore

matthewscottgordon committed:
Add GitHub Actions workflow to deploy to GitHub Pages

For use as staging environment.

matthewscottgordon committed:
Add on-push trigger for this branch in new GitHub action

Annoyingly, GitHub actions cannot be manually triggered until they've
been automatically triggered at least once.

matthewscottgordon committed:
Fix typo

matthewscottgordon committed:
Build on any PR but only deploy on main

tienhoah commented:
> Add GitHub Actions workflow to build and deploy to GitHub pages.
> 
> @tienhoah this is set up to build when there's a pull request and then deploy when it's merged to main. I'm guessing that's probably good for this project: main gets automatically deployed to a staging environment and deployment to production happens manually. Let me know if you want to do it differently.
> 
> @alexgleith I don't have the permissions needed to actually set up the deployment environment (`Branch "28-deploy-to-github-pages" is not allowed to deploy to github-pages due to environment protection rules.`) but I assume you can set that up and connect the custom subdomain?

@matthewscottgordon This sounds good to me.

alexgleith commented:
> actually set up the deployment environment

You don't want to deploy off branches/pull requests anyway, GH Pages is the production environment, so should be deploying off `main`, right?

Happy to set it up, of course, and we can set up the subdomain too. When do you want to go live with it?

matthewscottgordon committed:
Change when deployment workflow is triggered

Trigger build-and-deploy workflow when:

- A PR is created or updated (build only)
- A release is created (build and deploy main)
- The workflow is manually triggered (build always, deploy if on main)

matthewscottgordon commented:
@alexgleith, I talked to @tienhoah yesterday and he's going to keep Netlify as a test/staging environment and we'll use GitHub pages as the production development, as you said.

I updated the workflor trigger so that:

- Whenever a PR is created or updated, the build step runs but not the deployment step. This is just so we'll find out if the build step gets broken.
- Build and deployment will be triggered off of main whenever a new release is created in GitHub.
- The workflow can also be triggered manually. In this case, the build step always happens and the deploy step happens only if it's on `main`.

alexgleith commented:
Sounds good!
